RFCs/rfc-0004-fast-transactions.md

```diff
@@ -0,0 +1,77 @@
+# RFC-0004: Decreasing the minimum interval between SmartContract's Transactions from 3 minutes to 1 second with Recursive Proofs
+
+- **Intent**: Implement recursive proofs in SmartContract transactions to reduce the minimum transaction interval from 3 minutes to 1 second.
+- **Submitted by**: DFST ([Github](https://github.com/dfstio), [Twitter](https://twitter.com/dfst_io), Discord: dfst.io)
+- **Date**: Thursday, December 07, 2023
+
+## Abstract
+
+This RFC aims to improve the Mina Protocol's developer and user experience by enabling faster transaction processing in SmartContracts. Currently, a minimum interval of 3 minutes is required between SmartContract transactions that have preconditions and alter the state. This proposal intends to shorten this interval to 1 second, allowing rapid transaction execution with state preconditioning and changing.
+
+## Introduction
+
+Mina Protocol's SmartContract transactions, authenticated through proof, must meet these criteria:
+
+- Matching account update preconditions with the current account state.
+- Proof verification against the SmartContract's verification key.
+- Fee alignment with the market rate.
+
+Once verified, the transaction's inclusion in the block is assured, bypassing the need for blockchain-wide recalculations. Despite a known post-verification state, SmartContracts must wait about 3 minutes for state reflection in the new block before initiating another transaction.
+
+Current approaches to multiple transactions within a block include:
+
+- Non-use of preconditions and nonce manipulation (for example, in the case of the stateless contract).
+- Recursive proofs via ZkProgram for bundled transactions.
+- Using Actions and Reducer
+
+These methods, especially recursive proofs, pose significant entry barriers due to complexity and infrastructure demands and do not allow for many uses, as in the case of Actions and Reducer.
+
+**Problem**: The Mina blockchain's slow transaction pace (3 to 20 minutes) hinders rapid processing, contrasting with other blockchains like Tron (3 seconds) and Ethereum (12 seconds).
+
+**Relevance**: Accelerating transaction speed can greatly enhance the blockchain's Total Value Locked (TVL) and user experience, fostering Mina's ecosystem growth.
```

Comment:
Can you define who the core consumer of blockspace would be as the block-time becomes faster? I personally believe that the consumer of blockspace who uses actions + reducers or transactions on the L1 is likely very different from the consumer of blockspace who uses off-chain recursive proofs and ZkPrograms and submits a single state update. 

It would be great to outline why the idea for faster block-time is applicable, useful, and important to each different consumer persona.

Reply:
I would consider two types of customers to give the correct answer to your question: end customers and developers.

**The end customers** hardly know and do not care what is under the hood, but they do care about two things:
- Speed of sending the transaction
- Cost of sending the transactions
You can see how Tron TRC-20 tokens took the leading role in money transactions mainly because the transfers are fast and cheap, and Tron TVL is now 4 times higher than Polygon TVL, notwithstanding the fact of $2B investments into Polygon and many advanced ZK technologies integrated into Polygon.

It applies to all types of applications. The attention span of the people now is 8 seconds, and being able to deliver the result within those 8 seconds leads to a conversion increase. If the result is delivered after 8 seconds, many people have already forgotten what they would be doing and switched their attention to the next thing.

**The developers** are looking first to satisfy the end customers, so the above is of utmost importance for them. 

They have a wide range of tools available to them, and one of the tools they have used in recent years, as I see from my experience, is not to use the blockchain at all and base the design on old architecture to meet their business objectives, so the first factor we are speaking about is the adoption of the blockchain technology. 

The other tools as L2, recursive proofs, Actions/Reducer are available for them, and they will choose the tool on the
- Easy to use
- Ability to meet the objectives
- Cost
- SLA levels

In this sense, all those solutions are complementary. Actions/Reducer is easy to use and does not require external infrastructure and additional cost, but you cannot update Merkle Map with it.

Recursive proofs are very powerful but require a significant CPU and memory. Fast calculation and merging of even 32 proofs in a reasonable time requires the computer to have the parameters that exist only in a cloud/server environment. Setting up such an environment is difficult for the developer who just came to the ecosystem.

L2 is probably the best solution, but the questions of data availability, messaging, and asset transfer between L1-L2 are always complicating the development.

The experienced developer can handle all this, but if we are speaking about a junior developer, he is looking to implement his idea based on a slight modification of the example from the tutorial and is not looking to integrate with L2 or cloud infrastructure except when this integration is very simple and already included into example))

The focus of this proposal is to provide capabilities equal to L2 but without the need to change the tutorial code or do some integrations. In some sense, you can consider this proposal as a “light” L2 with a life span of 3 minutes that, at the end of its life, as a new block comes, transfers all the assets and state to the L1 and disappears.

To summarize, this proposal is essential for applications that have to make transactions fast and cheap without significant CPU and memory requirements.

The example of such applications are:
- Games
- Trading (and to withdraw safety, you still need to wait for a block, but trading can be done within the block)
- Putting live data into the blockchain in real-time (oracles for quotes, legally important audio, other real-time data streams)
- Real-time voting in the meetings
- Any transactions where identity should be present to the zkApp and relied on to make the transaction, such as including the person in a white list before making the transaction.

Reply:
![speed](https://github.com/MinaFoundation/Core-Grants/assets/110384358/7d34df83-3496-434d-b088-5fe0715d1421)

Reply:
You raise many fantastic points here. Please include this in the RFC's markdown!

In my opinion, the primary challenge to scaling Web3 applications for many users is eliminating the system complexity that manifests in user experiences. Congestion, high fees, and long waiting times are just a few of the complexities of a blockchain that many users experience in a Web3 application. One way to mask these complexities is with higher TPS or similar workarounds, but these seldom solve the problems, only delaying them. Still, these complexities do not exist in end-users mental models of the Web3 application they're interfacing with, and they arise in their interaction with Web2 applications. Ephemeral zkVMs are one interesting way to address these challenges with the first-in-class features Mina has.

I'm keen to see if others have thoughts in this direction too.

Reply:
I've included more detailed opinion about the Action/Reducer. It is a great tool and I've also included into RFC some ideas on how it can be enhanced with a disclaimer that it is not a scope of the current RFC. There is also good discussion on discord on Actions and Reducer:
https://discord.com/channels/484437221055922177/1173407599123767346

I've amended the title and introduction in accordance with your recommendation. Thank you for it, I really like your wording!